packages/selected-network-controller/src/SelectedNetworkController.ts

```diff
@@ -0,0 +1,166 @@
+import type { RestrictedControllerMessenger } from '@metamask/base-controller';
+import { BaseControllerV2 } from '@metamask/base-controller';
+import type {
+  NetworkClientId,
+  NetworkControllerStateChangeEvent,
+  NetworkState,
+} from '@metamask/network-controller';
+import type { Patch } from 'immer';
+
+const controllerName = 'SelectedNetworkController';
+
+const stateMetadata = {
+  domains: { persist: true, anonymous: false },
+  perDomainNetwork: { persist: true, anonymous: false },
+};
+
+const getDefaultState = () => ({
+  domains: {},
+  perDomainNetwork: false,
+});
+
+type Domain = string;
+
+const METAMASK_DOMAIN = 'metamask' as const;
+
+export const SelectedNetworkControllerActionTypes = {
+  getState: `${controllerName}:getState` as const,
+  getNetworkClientIdForDomain:
+    `${controllerName}:getNetworkClientIdForDomain` as const,
+  setNetworkClientIdForDomain:
+    `${controllerName}:setNetworkClientIdForDomain` as const,
+};
+
+export const SelectedNetworkControllerEventTypes = {
+  stateChange: `${controllerName}:stateChange` as const,
+};
+
+export type SelectedNetworkControllerState = {
+  domains: Record<Domain, NetworkClientId>;
+  // Feature flag to start returning networkClientId based on the domain.
+  // when the flag is false, the 'metamask' domain will always be used.
+  // defaults to false
+  perDomainNetwork: boolean;
+};
+
+export type SelectedNetworkControllerStateChangeEvent = {
+  type: typeof SelectedNetworkControllerEventTypes.stateChange;
+  payload: [SelectedNetworkControllerState, Patch[]];
+};
+
+export type SelectedNetworkControllerGetSelectedNetworkStateAction = {
+  type: typeof SelectedNetworkControllerActionTypes.getState;
+  handler: () => SelectedNetworkControllerState;
+};
+
+export type SelectedNetworkControllerGetNetworkClientIdForDomainAction = {
+  type: typeof SelectedNetworkControllerActionTypes.getNetworkClientIdForDomain;
+  handler: (domain: string) => NetworkClientId;
+};
+
+export type SelectedNetworkControllerSetNetworkClientIdForDomainAction = {
+  type: typeof SelectedNetworkControllerActionTypes.setNetworkClientIdForDomain;
+  handler: (domain: string, NetworkClientId: NetworkClientId) => void;
+};
+
+export type SelectedNetworkControllerActions =
```

Comment:
`SelectedNetworkControllerAction` singular?

Reply:
since its a collection of actions, I think plural makes more sense.